Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp

```diff
@@ -386,64 +392,115 @@ String WebsiteDataStore::migrateMediaKeysStorageIfNecessary(const String& direct
+    if (!directories.modelElementCacheDirectory.isEmpty())
+        result.modelElementCacheDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(directories.modelElementCacheDirectory);
 #endif
-    if (!configuration->searchFieldHistoryDirectory().isEmpty())
-        resolvedConfiguration->setSearchFieldHistoryDirectory(resolveAndCreateReadWriteDirectoryForSandboxExtension(configuration->searchFieldHistoryDirectory()));
 
-    // Resolve file paths.
-    if (!configuration->cookieStorageFile().isEmpty()) {
-        auto resolvedCookieDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(FileSystem::parentPath(configuration->cookieStorageFile()));
-        resolvedConfiguration->setCookieStorageFile(FileSystem::pathByAppendingComponent(resolvedCookieDirectory, FileSystem::pathFileName(configuration->cookieStorageFile())));
+    if (!directories.cookieStorageFile.isEmpty()) {
+        auto resolvedCookieDirectory = resolveAndCreateReadWriteDirectoryForSandboxExtension(FileSystem::parentPath(directories.cookieStorageFile));
+        result.cookieStorageFile = (FileSystem::pathByAppendingComponent(resolvedCookieDirectory, FileSystem::pathFileName(directories.cookieStorageFile)));
     }
 
-    // Default paths of WebsiteDataStore created with identifier are not under caches or tmp directory,
-    // so we need to explicitly exclude them from backup.
-    if (configuration->identifier()) {
+    result.webSQLDatabaseDirectory = directories.webSQLDatabaseDirectory;
+    result.applicationCacheDirectory = directories.applicationCacheDirectory;
+
+    return result;
+}
+
+WebsiteDataStoreConfiguration::Directories WebsiteDataStore::resolvedDirectories()
+{
+    Locker resolveLocker { m_resolveLock };
+    return m_resolvedConfiguration->directories();
+}
+
+void WebsiteDataStore::dispatchResolveDirectories()
+{
+    protectedQueue()->dispatch([this, protectedThis = Ref { *this }, directories = m_configuration->directories().isolatedCopy()]() mutable {
+        auto resolvedDirectories = resolveDirectories(directories);
+        Locker resolveLocker { m_resolveLock };
+        m_resolvedConfiguration->setDirectories(resolvedDirectories.isolatedCopy());
+        m_resolveState = ResolveState::Resolved;
+        m_resolveCondition.notifyOne();
+        RunLoop::main().dispatch([protectedThis = WTFMove(protectedThis)] { });
+    });
+}
+
+void WebsiteDataStore::resolveDirectoriesIfNecessary()
+{
+    ASSERT(RunLoop::isMain());
+
+    Locker resolveLocker { m_resolveLock };
+    if (m_resolveState == ResolveState::Resolved)
+        return;
+
+    if (m_resolveState == ResolveState::Unresolved) {
+        dispatchResolveDirectories();
+        m_resolveState = ResolveState::Resolving;
+    }
+
+    while (m_resolveState != ResolveState::Resolved)
+        m_resolveCondition.wait(m_resolveLock);
```

Comment:
Hanging the high priority main thread on a low priority background thread doesn't seem like a good idea to me. Was this perf tested? This adds a lot of complexity and it isn't a clear win (since there is a code to dispatching back and forth between threads, and we'll end up hanging the main thread on the filesystem operations anyway if need the resolved directories very shortly after creating the data store (which likely we will). Not to mention the thread priority issue (but that's easily fixable.

Reply:
> Was this perf tested?

Looks like no performance impact on PLT; see link in the radar.

> we'll end up hanging the main thread on the filesystem operations anyway if need the resolved directories very shortly after creating the data store (which likely we will)

It depends on the use case. In the case where client initializes the data store but not use it for loading yet (e.g. creating a global data store when browser launches, like prewarming web process), this will be a perf improvement since the directories will be ready before using the data store, also main thread could keep responsive for user input. I noticed the file operations are slow when I investigated rdar://114388755 as they showed up in the profile.

As for the thread priority issue, I didn't notice much of the difference, probably because the system is not busy enough. But if you think it's better to bump the priority of the work queue to ensure main thread is unblocked, I could make the change.

Reply:
> As for the thread priority issue, I didn't notice much of the difference, probably because the system is not busy enough. But if you think it's better to bump the priority of the work queue to ensure main thread is unblocked, I could make the change.

I don't think the PR should be merged without this. It is not OK to block the main thread on a lower priority thread. It partly defeats the purpose of this optimization.

Note that I don't know if ALL WKWebsiteDataStore background operations should be done on a high priority thread. However, it seems clear the directory resolving does, given that the main thread may block on the operation being done. I didn't check what other operations we do on that work queue but it's possible we'll need 2 work queues for e.g.

Reply:
Two work queues with different priorities seems fine; I will update the patch